### PR TITLE
Create additional integration test cases

### PR DESCRIPTION
**Pull request by querti**

### CODE REVIEW

**Comment by rbikar** [comment]
run tests

File: tests/integration/test_association.py [hunk 1]
```diff
@@ -241,6 +254,39 @@ def assert_empty_repos(cfg):
...
+    '''
+    Test if modulemd_defaults are copied to ubi repository.
+    '''
+    cfg = load_ubiconfig('associate-md.yaml')
+    rpm_repos = get_repos_from_cs(cfg.content_sets.rpm.output, skip_dot_version=True)
+    for repo in rpm_repos:
+        repo_url = repo_url = get_repo_url(repo['url'])
+        break
+    modules = query_repo_modules(None, 'ubi', repo_url, full_data=True)
+    print modules
```
Comment (rbikar) [inline tests/integration/test_association.py]: Not python3 compatible.

see travis error:

> /opt/python/3.5.6/lib/python3.5/ast.py:35: in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
E     File "/home/travis/build/release-engineering/ubi-population-tool/tests/integration/test_association.py", line 268
E       print modules
E                   ^
E   SyntaxError: Missing parentheses in call to 'print'

**Comment by querti** [comment]
run tests

File: tests/integration/test_association.py (hunk 1 above)
Comment (querti) [reply tests/integration/test_association.py]: Yes, I have removed it

**Comment by querti** [comment]
@emilyzheng Can you also review pls?

**Comment by querti** [comment]
run tests

**Comment by rbikar** [comment]
run tests

**Comment by rbikar** [comment]
run tests

File: tests/integration/test_association.py [hunk 2]
```diff
@@ -241,6 +254,39 @@ def assert_empty_repos(cfg):
...
+    for repo in rpm_repos:
+        repo_url = repo_url = get_repo_url(repo['url'])
+        break
+    modules = query_repo_modules(None, 'ubi', repo_url, full_data=True)
+
+    assert len(modules) == 2, 'Unexpected repo modules found.'
+    separated_modules = [(module[:module.find(' ')], module[module.find(' ') + 1:])
+                         for module in modules]
+    
+    for module in separated_modules:
```
Comment (rbikar) [inline tests/integration/test_association.py]: Just a thought here.
This serves as a validation of testing data.
I think in future, it should be better to reset testing repositories/units on pulp server before testing to some baseline state.
This can get messed easily and testing then will be unreliable.

This also applies for other parts of code, but this should solved in different task since it might require non trivial effort.

File: tests/integration/test_association.py [hunk 3]
```diff
@@ -241,6 +254,39 @@ def assert_empty_repos(cfg):
...
+        elif module[0] == 'perl-FCGI':
+            assert '[d]' in module[1], 'Module perl-FCGI should have defaults.'
+        else:
+            raise AssertionError('Unknown module in repo.')
+    
+    run_ubipop_tool('associate-md.yaml')
+
+    modules = query_repo_modules(None, 'ubi', repo_url, full_data=True)
+    assert len(modules) == 1, 'Unexpected repo modules found.'
+    separated_module = (modules[0][:modules[0].find(' ')], modules[0][modules[0].find(' ') + 1:])
```
Comment (rbikar) [inline tests/integration/test_association.py]: Looks to me a bit crude parsing of yum list module output.
Please consider moving this to some helper method which parse yum output. It's also use on line 270

File: tests/integration/test_association.py [hunk 4]
```diff
@@ -241,6 +254,39 @@ def assert_empty_repos(cfg):
...
+    assert separated_module[0] == 'httpd', 'Unknown module in repo.'
+    assert '[d]' in separated_module[1], 'Module httpd should have defaults.'
```
Comment (rbikar) [inline tests/integration/test_association.py]: Shouldn't this also assert which profile is default? Not just that there some profile was set as default.

File: tests/integration/test_association.py [hunk 5]
```diff
@@ -453,3 +499,109 @@ def assert_mds(mds_included, mds_found):
...
+               'Unexpected rpm in the debuginfo content set.'
+
+    logs = run_ubipop_tool('associate-md-pkg.yaml', capture_log=True)
+
+    rpm_repos = get_repos_from_cs(cfg.content_sets.rpm.output, skip_dot_version=True)
+    for repo in rpm_repos:
+        repo_url = repo_url = get_repo_url(repo['url'])
+        break
+    repo_rpms = query_repo_rpms(None, 'ubi', repo_url)
+    assert expected1[0] in repo_rpms and expected1[1] in repo_rpms, \
```
Comment (rbikar) [inline tests/integration/test_association.py]: It's kind of difficult to review this because I don't how module's artifacts list looks...

File: tests/integration/test_association.py [hunk 6]
```diff
@@ -453,3 +499,109 @@ def assert_mds(mds_included, mds_found):
...
+           'Package in module profile has not been added to ubi repo.'
+    assert expected1[2] in repo_rpms, \
+           'Package in module artifacts has not been added to ubi repo.'
+    assert expected1[3] in repo_rpms, \
+           'Package in module artifacts, which is also excluded has not been added to ubi repo.'
+
+    debug_repos = get_repos_from_cs(cfg.content_sets.debuginfo.output, skip_dot_version=True)
+    for repo in debug_repos:
+        repo_url = repo_url = get_repo_url(repo['url'])
```
Comment (rbikar) [inline tests/integration/test_association.py]: What's the intention here?
Also on 575, 549 and other lines.

File: tests/integration/test_association.py [hunk 7]
```diff
@@ -453,3 +499,109 @@ def assert_mds(mds_included, mds_found):
...
+    '''
+    Test if packages defined in a module profile and artifacts have been added to ubi repo.
+    '''
+    cfg = load_ubiconfig('associate-md-pkg.yaml')
+    rpm_repos = get_repos_from_cs(cfg.content_sets.rpm.output, skip_dot_version=True)
+    for repo in rpm_repos:
+        repo_url = repo_url = get_repo_url(repo['url'])
+        break
+    repo_rpms = query_repo_rpms(None, 'ubi', repo_url)
+    expected1 = ['httpd-2.4.6-88.el7.x86_64.rpm', 'httpd-tools-2.4.6-88.el7.x86_64.rpm',
```
Comment (rbikar) [inline tests/integration/test_association.py]: What's the deal with expected1[1] item? Doesn't seem tested in any way.

File: tests/integration/test_association.py (hunk 2 above)
Comment (querti) [reply tests/integration/test_association.py]: Currently, the data is only reset before the whole test suite is run. It should be possible to reset it between each test, even though it would make the runtime significantly higher. Another solution might be to prepare exclusive data for each test, so that tests wouldn't interfere with each other. For now, the suite works with the existing data, but additional data would probably be needed in the future if additional test cases are added.

**Comment by querti** [comment]
run tests

File: tests/integration/test_association.py [hunk 8]
```diff
@@ -453,3 +506,109 @@ def assert_mds(mds_included, mds_found):
...
+    '''
+    cfg = load_ubiconfig('associate-pkg-multi-arch.yaml')
+    rpm_repos = get_repos_from_cs(cfg.content_sets.rpm.output, skip_dot_version=True)
+    for repo in rpm_repos:
+        repo_url = repo_url = get_repo_url(repo['url'])
+        break
+    repo_rpms = query_repo_rpms(None, 'ubi', repo_url)
+    assert all('glibc' not in rpm for rpm in repo_rpms), \
+           'No glibc rpms should be in the output repo.'
+
```
Comment (midnightercz) [inline tests/integration/test_association.py]: I wonder if you should also add here to test check input data. That you make it easier to distinguish between failure due missing input data and wrong ubi-pop functionality.

File: tests/integration/test_association.py (hunk 3 above)
Comment (querti) [reply tests/integration/test_association.py]: I have added a helper function

File: tests/integration/test_association.py (hunk 4 above)
Comment (querti) [reply tests/integration/test_association.py]: The test should now check for which profile is default

**Comment by querti** [comment]
run tests

File: tests/integration/test_association.py (hunk 2 above)
Comment (querti) [reply tests/integration/test_association.py]: changed

File: tests/integration/test_association.py (hunk 6 above)
Comment (querti) [reply tests/integration/test_association.py]: Copy-paste gone wrong. I have fixed it

File: tests/integration/test_association.py (hunk 7 above)
Comment (querti) [reply tests/integration/test_association.py]: It is tested.

File: tests/integration/test_association.py [hunk 9]
```diff
@@ -241,6 +260,40 @@ def assert_empty_repos(cfg):
...
 
+@pytest.mark.skipif(INTEGRATION_NOT_SETUP, reason='Integration test is not set up.')
+def test_modulemd_defaults_unit_copy():
+    '''
+    Test if modulemd_defaults are copied to ubi repository.
+    '''
+    cfg = load_ubiconfig('associate-md.yaml')
+    rpm_repos = get_repos_from_cs(cfg.content_sets.rpm.output, skip_dot_version=True)
+    for repo in rpm_repos:
+        repo_url = repo_url = get_repo_url(repo['url'])
```
[the remark this answers is not in the document]
Comment (querti) [reply tests/integration/test_association.py]: It is a generator.

File: tests/integration/test_association.py [hunk 10]
```diff
@@ -241,6 +260,40 @@ def assert_empty_repos(cfg):
...
+                   'Module perl-FCGI should have common profile as default.'
+        else:
+            raise AssertionError('Unknown module in repo.')
+    
+    run_ubipop_tool('associate-md.yaml')
+
+    modules = query_repo_modules(None, 'ubi', repo_url, full_data=True)
+    assert len(modules) == 1, 'Unexpected repo modules found.'
+    separated_module = separate_modules(modules)[0]
+    assert separated_module[0] == 'httpd', 'Unknown module in repo.'
```
[the remark this answers is not in the document]
Comment (querti) [reply tests/integration/test_association.py]: Fixed

File: tests/integration/test_association.py [hunk 11]
```diff
@@ -453,3 +506,109 @@ def assert_mds(mds_included, mds_found):
...
+           'No glibc rpms should be in the output repo.'
+
+    run_ubipop_tool('associate-pkg-multi-arch.yaml')
+
+    repo_rpms = query_repo_rpms(None, 'ubi', repo_url)
+    found_archs = []
+    for rpm in repo_rpms:
+        if 'glibc' in rpm:
+            found_archs.append(rpm.split('.')[-2])
+    assert all(arch in found_archs for arch in ['i686', 'x86_64']), \
```
[the remark this answers is not in the document]
Comment (querti) [reply tests/integration/test_association.py]: Changed.

File: tests/integration/test_association.py (hunk 8 above)
Comment (querti) [reply tests/integration/test_association.py]: Jenkins job restores all of the data before running the test suite. Also, Emily told me that I would either need to use the pa_tool or parse the database directly to check the input data. Is setting this up for such an edge case worth it?

**Comment by querti** [comment]
run tests
